File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.py
sign = ["+", "-"]
l = []

def check(t) :
    global target_num
    if sum(t) == target_num :
        return True
    else : return False

def dfs(count) :
    global limit
    global answer
    global list_numbers
    if count == limit :
        if check(l) :
            answer += 1
        return
    
    for i in range(len(sign)) :
        l.append(int(sign[i] + str(list_numbers[count])))
        dfs(count + 1)
        l.pop()
    
def solution(numbers, target):
    global limit
    global answer
    global list_numbers
    list_numbers = numbers
    global target_num
    target_num = target
    
    limit = len(numbers)
    answer = 0
    
    dfs(0)
    
    
    return answer

File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.java
import java.util.*;

class Solution {
    public static int solution(int[] numbers, int target) {
        int[] answer = {0};
        String[] operation = new String[]{"+", "-"};
        dfs(operation, numbers, target, 0, new ArrayList<>(), answer);
        return answer[0];
    }

    public static void dfs(String[] operation, int[] numbers, int target, int count, List<Integer> now, int[] answer) {
        if (count == numbers.length) {
            int sum = now.stream()
                    .mapToInt(Integer::intValue)
                    .sum();
            if (sum == target) {
                answer[0]++;
            }
            return;
        }

        for (int i = 0; i < operation.length; i++) {
            int number = numbers[count];
            if (operation[i].equals("-")) {
                number = -number;
            }

            now.add(number);
            dfs(operation, numbers, target, count + 1, now, answer);
            now.remove(now.size() - 1);
        }
    }
}